[Profiler] Propagate metadata into `Engine::evaluate_function` event.

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #77697
* __->__ #77696
* #77695
* #77694

https://github.com/pytorch/pytorch/pull/63619 added a RECORD_FUNCTION guard to make calls to `Engine::evaluate_function` visible regardless of the underlying op. While useful, this creates a call that looks like a forward call that somewhat complicates stitching forward and backward ops. I don't want to add complexity (and therefore work) on the hot path; instead it's fairly straightforward to stitch things back together in post. This PR simply propagates sequence number and forward tid info up to the `evaluate_function` event.

Differential Revision: [D36302562](https://our.internmc.facebook.com/intern/diff/D36302562/)

Commit:
[Profiler] Propagate metadata into `Engine::evaluate_function` event.

https://github.com/pytorch/pytorch/pull/63619 added a RECORD_FUNCTION guard to make calls to `Engine::evaluate_function` visible regardless of the underlying op. While useful, this creates a call that looks like a forward call that somewhat complicates stitching forward and backward ops. I don't want to add complexity (and therefore work) on the hot path; instead it's fairly straightforward to stitch things back together in post. This PR simply propagates sequence number and forward tid info up to the `evaluate_function` event.

Differential Revision: [D36302562](https://our.internmc.facebook.com/intern/diff/D36302562/)

[ghstack-poisoned]

aaronenyeshi reviewed (approved):
Reviewed internally, LGTM!

Commit:
Update on "[Profiler] Propagate metadata into `Engine::evaluate_function` event."


https://github.com/pytorch/pytorch/pull/63619 added a RECORD_FUNCTION guard to make calls to `Engine::evaluate_function` visible regardless of the underlying op. While useful, this creates a call that looks like a forward call that somewhat complicates stitching forward and backward ops. I don't want to add complexity (and therefore work) on the hot path; instead it's fairly straightforward to stitch things back together in post. This PR simply propagates sequence number and forward tid info up to the `evaluate_function` event.

Differential Revision: [D36302562](https://our.internmc.facebook.com/intern/diff/D36302562/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Propagate metadata into `Engine::evaluate_function` event."


https://github.com/pytorch/pytorch/pull/63619 added a RECORD_FUNCTION guard to make calls to `Engine::evaluate_function` visible regardless of the underlying op. While useful, this creates a call that looks like a forward call that somewhat complicates stitching forward and backward ops. I don't want to add complexity (and therefore work) on the hot path; instead it's fairly straightforward to stitch things back together in post. This PR simply propagates sequence number and forward tid info up to the `evaluate_function` event.

Differential Revision: [D36302562](https://our.internmc.facebook.com/intern/diff/D36302562/)

[ghstack-poisoned]
